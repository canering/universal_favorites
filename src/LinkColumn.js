import React, {useState} from "react";
import PropTypes from "prop-types";
import Link from "./Link";

function LinkColumn(props) {
    const [draggingId, setDraggingId] = useState(0);

    return (
        <div>
            {props.links.map((item, id) => (
                <div key={item.id} onContextMenu={(event) => props.showContextMenu(event, id)}>
                    <Link 
                        item={item} id={id} openFolder={props.openFolder} draggingId={draggingId} 
                        dropElement={props.dropElement} saveDraggingId={(id) => setDraggingId(id)}
                    />
                </div>)
            )}
        </div>);
}

LinkColumn.propTypes = {
    links:PropTypes.array,
    openFolder:PropTypes.func,
    dropElement:PropTypes.func,
    showContextMenu:PropTypes.func
};

export default LinkColumn;