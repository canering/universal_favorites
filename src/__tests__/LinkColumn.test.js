import React from "react";
import {shallow, mount, configure} from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import LinkColumn from "../LinkColumn";
import chrome from "sinon-chrome";
import toJson from "enzyme-to-json";
import TestBackend from "react-dnd-test-backend";
import { DndProvider } from "react-dnd";


window.chrome = chrome;

configure({adapter: new Adapter()});

const props = {
    links: [
        {type:"folder", label:"folder", id:0, data:[
            {type:"link", link:"http://example.com", label:"example.com", id:2},
            {type:"link", link:"http://example2.com", label:"example2.com", id:3}
        ]},
        {type:"link", link:"http://example.com", label:"example.com", id:1},
        {type:"link", link:"http://example.com", label:"example.com", id:4}
    ],
    changeLinks: jest.fn(),
    openFolder: jest.fn(),
    showEdit: jest.fn(),
    deleteLink: jest.fn(),
    tempChangeLinks: jest.fn()
};

var component;

beforeEach(() => {
    component = shallow(<LinkColumn {...props} />);
});

afterEach(() => {
    jest.clearAllMocks();
});

test("link column snapshot test", async () => {
    let tree = toJson(component);
    expect(tree).toMatchSnapshot();

    let links = component.find(".LinkColumnDiv");
    links.at(0).simulate("contextmenu", {clientX:20, preventDefault: jest.fn()});
    component.update();
    
    tree = toJson(component);
    expect(tree).toMatchSnapshot();
});

test("when edit is clicked in context menu then showEdit should be called", () => {
    let links = component.find(".LinkColumnDiv");
    links.at(0).simulate("contextmenu", {clientX:20, preventDefault: jest.fn()});

    let contextMenu = component.find("#ContextMenu");
    contextMenu.children().at(0).prop("onClick")({});

    expect(props.showEdit).toHaveBeenCalledTimes(1);
    expect(props.showEdit).toHaveBeenCalledWith(props.links[0]);
});

test("when delete is clicked in context menu then deleteLink should be called", () => {
    let links = component.find(".LinkColumnDiv");
    links.at(0).simulate("contextmenu", {clientX:20, preventDefault: jest.fn()});

    let contextMenu = component.find("#ContextMenu");
    contextMenu.children().at(1).prop("onClick")({});

    expect(props.deleteLink).toHaveBeenCalledTimes(1);
    expect(props.deleteLink).toHaveBeenCalledWith(props.links[0].id);
});

test("when Open All is clicked in a folder's context menu then all links inside should be opened", () => {
    let links = component.find(".LinkColumnDiv");
    links.at(0).simulate("contextmenu", {clientX:20, preventDefault: jest.fn()});

    let contextMenu = component.find("#ContextMenu");
    contextMenu.children().at(2).prop("onClick")({});

    expect(chrome.tabs.create.called).toBe(true);
    expect(chrome.tabs.create.callCount).toBe(2);
    expect(chrome.tabs.create.calledWith({url:props.links[0].data[0].link, active:false}));
    expect(chrome.tabs.create.calledWith({url:props.links[0].data[1].link, active:false}));
});

test("when close is clicked in context menu then context menu should be hidden", () => {
    let links = component.find(".LinkColumnDiv");
    links.at(0).simulate("contextmenu", {clientX:20, preventDefault: jest.fn()});

    let contextMenu = component.find("#ContextMenu");
    expect(contextMenu.prop("hidden")).toBe(false);

    contextMenu.children().at(3).prop("onClick")({});

    contextMenu = component.find("#ContextMenu");
    expect(contextMenu.prop("hidden")).toBe(true);
});

test("when shift clicking appropriate links should be marked for App to update", () => {
    let mountedComponent = mount(<DndProvider backend={TestBackend}><LinkColumn {...props} /></DndProvider>);

    let expectedResult = props.links.map(a => ({...a}));
    expectedResult[0].clicked = true;

    let items = mountedComponent.find(".menu-item");
    items.at(0).prop("onClick")({shiftKey:true});

    expect(props.tempChangeLinks).toHaveBeenCalledTimes(1);
    expect(props.tempChangeLinks).toHaveBeenCalledWith(expectedResult);

    let updatedProps = Object.assign({}, props, {links:expectedResult});
    mountedComponent = mount(<DndProvider backend={TestBackend}><LinkColumn {...updatedProps} /></DndProvider>);
    items = mountedComponent.find(".menu-item");

    items.at(2).prop("onClick")({shiftKey:true});

    expectedResult[1].clicked = true;
    expectedResult[2].clicked = true;

    expect(props.tempChangeLinks).toHaveBeenCalledTimes(2);
    expect(props.tempChangeLinks).toHaveBeenLastCalledWith(expectedResult);
});

test("when ctrl clicking appriate link should be marked for app to update", () => {
    let mountedComponent = mount(<DndProvider backend={TestBackend}><LinkColumn {...props} /></DndProvider>);

    let expectedResult = props.links.map(a => ({...a}));
    expectedResult[0].clicked = true;

    let items = mountedComponent.find(".menu-item");
    items.at(0).prop("onClick")({ctrlKey:true});

    expect(props.tempChangeLinks).toHaveBeenCalledTimes(1);
    expect(props.tempChangeLinks).toHaveBeenCalledWith(expectedResult);
});