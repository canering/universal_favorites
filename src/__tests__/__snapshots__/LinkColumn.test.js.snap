// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`link column snapshot test 1`] = `
<div
  style={
    Object {
      "position": "relative",
    }
  }
>
  <div
    className="LinkColumnDiv"
    key="0"
    onContextMenu={[Function]}
  >
    <Link
      draggingId={0}
      dropElement={[Function]}
      id={0}
      item={
        Object {
          "data": Array [
            Object {
              "id": 2,
              "label": "example.com",
              "link": "http://example.com",
              "type": "link",
            },
            Object {
              "id": 3,
              "label": "example2.com",
              "link": "http://example2.com",
              "type": "link",
            },
          ],
          "id": 0,
          "label": "folder",
          "type": "folder",
        }
      }
      openFolder={[MockFunction]}
      saveDraggingId={[Function]}
      setClicked={[Function]}
      setShiftClicked={[Function]}
    />
  </div>
  <div
    className="LinkColumnDiv"
    key="1"
    onContextMenu={[Function]}
  >
    <Link
      draggingId={0}
      dropElement={[Function]}
      id={1}
      item={
        Object {
          "id": 1,
          "label": "example.com",
          "link": "http://example.com",
          "type": "link",
        }
      }
      openFolder={[MockFunction]}
      saveDraggingId={[Function]}
      setClicked={[Function]}
      setShiftClicked={[Function]}
    />
  </div>
  <div
    className="LinkColumnDiv"
    key="4"
    onContextMenu={[Function]}
  >
    <Link
      draggingId={0}
      dropElement={[Function]}
      id={2}
      item={
        Object {
          "id": 4,
          "label": "example.com",
          "link": "http://example.com",
          "type": "link",
        }
      }
      openFolder={[MockFunction]}
      saveDraggingId={[Function]}
      setClicked={[Function]}
      setShiftClicked={[Function]}
    />
  </div>
  <div
    hidden={true}
    id="ContextMenu"
    style={
      Object {
        "left": 0,
        "top": 0,
      }
    }
  >
    <span
      onClick={[Function]}
    >
      Edit
    </span>
    <span
      onClick={[Function]}
    >
      Delete
    </span>
    <span
      onClick={[Function]}
    >
      Close
    </span>
  </div>
</div>
`;

exports[`link column snapshot test 2`] = `
<div
  style={
    Object {
      "position": "relative",
    }
  }
>
  <div
    className="LinkColumnDiv"
    key="0"
    onContextMenu={[Function]}
  >
    <Link
      draggingId={0}
      dropElement={[Function]}
      id={0}
      item={
        Object {
          "data": Array [
            Object {
              "id": 2,
              "label": "example.com",
              "link": "http://example.com",
              "type": "link",
            },
            Object {
              "id": 3,
              "label": "example2.com",
              "link": "http://example2.com",
              "type": "link",
            },
          ],
          "id": 0,
          "label": "folder",
          "type": "folder",
        }
      }
      openFolder={[MockFunction]}
      saveDraggingId={[Function]}
      setClicked={[Function]}
      setShiftClicked={[Function]}
    />
  </div>
  <div
    className="LinkColumnDiv"
    key="1"
    onContextMenu={[Function]}
  >
    <Link
      draggingId={0}
      dropElement={[Function]}
      id={1}
      item={
        Object {
          "id": 1,
          "label": "example.com",
          "link": "http://example.com",
          "type": "link",
        }
      }
      openFolder={[MockFunction]}
      saveDraggingId={[Function]}
      setClicked={[Function]}
      setShiftClicked={[Function]}
    />
  </div>
  <div
    className="LinkColumnDiv"
    key="4"
    onContextMenu={[Function]}
  >
    <Link
      draggingId={0}
      dropElement={[Function]}
      id={2}
      item={
        Object {
          "id": 4,
          "label": "example.com",
          "link": "http://example.com",
          "type": "link",
        }
      }
      openFolder={[MockFunction]}
      saveDraggingId={[Function]}
      setClicked={[Function]}
      setShiftClicked={[Function]}
    />
  </div>
  <div
    hidden={false}
    id="ContextMenu"
    style={
      Object {
        "left": 20,
        "top": 0,
      }
    }
  >
    <span
      onClick={[Function]}
    >
      Edit
    </span>
    <span
      onClick={[Function]}
    >
      Delete
    </span>
    <span
      onClick={[Function]}
    >
      Open All
    </span>
    <span
      onClick={[Function]}
    >
      Close
    </span>
  </div>
</div>
`;
